crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+        get_binding_stmt(group)
+    };
+
+    for append in appends.iter().skip(1) {
+        let group = get_or_add(&mut map, append);
+        let binding = get_binding_stmt(group);
+        if binding != last_binding {
+            // If the group is not brand new, and the previous group was different,
+            // we should mark it as "non-consecutive".
+            if group.appends.len() > 1 {
+                group.is_consecutive = false;
+            }
+            last_binding = binding;
+        }
+    }
+
+    map.into_values()
+        // Groups with just one element are not interesting for us.
+        .filter(|group| group.appends.len() > 1)
+        .collect()
+}
+
+#[inline]
+fn get_binding_stmt(group: &AppendGroup) -> NodeId {
+    group.appends.first().unwrap().binding.source.unwrap()
+}
+
+#[inline]
+fn get_or_add<'a, 'b>(
+    map: &'b mut HashMap<NodeId, AppendGroup<'a>>,
+    append: &Append<'a>,
+) -> &'b mut AppendGroup<'a> {
+    let binding_stmt = append.binding.source.unwrap();
+    let group = map.entry(binding_stmt).or_insert(AppendGroup {
+        appends: vec![],
+        is_consecutive: true,
+    });
+    group.appends.push(append.clone());
+    group
+}
+
+fn make_suggestion(group: &AppendGroup, generator: Generator) -> String {
+    let appends = &group.appends;
+
+    assert!(!appends.is_empty());
+    let first = appends.first().unwrap();
+
+    assert!(appends
+        .iter()
+        .all(|append| append.binding.source == first.binding.source));
+
+    // Here we construct `var.extend((elt1, elt2, ..., eltN))
+    //
+    // Each eltK comes from an individual `var.append(eltK)`.
+    let elts: Vec<ast::Expr> = appends
+        .iter()
+        .map(|append| append.argument.clone())
+        .collect();
+    // Join all elements into a tuple: `(elt1, elt2, ..., eltN)`
+    let tuple = ast::ExprTuple {
+        elts,
+        ctx: ast::ExprContext::Load,
+        range: TextRange::default(),
+    };
+    // Make `var.extend`.
+    // NOTE: receiver is the same for all appends and that's why we can take the first.
+    let attr = ast::ExprAttribute {
+        value: Box::new(first.receiver.clone().into()),
+        attr: ast::Identifier::new("extend".to_string(), TextRange::default()),
+        ctx: ast::ExprContext::Load,
+        range: TextRange::default(),
+    };
+    // Make the actual call `var.extend((elt1, elt2, ..., eltN))`
+    let call = ast::ExprCall {
+        func: Box::new(attr.into()),
+        arguments: ast::Arguments {
+            args: vec![tuple.into()],
+            keywords: vec![],
+            range: TextRange::default(),
+        },
+        range: TextRange::default(),
+    };
+    // And finally, turn it into a statement.
+    let stmt = ast::StmtExpr {
+        value: Box::new(call.into()),
+        range: TextRange::default(),
+    };
+    generator.stmt(&stmt.into())
+}
+
+fn match_append<'a>(checker: &'a Checker, stmt: &'a Stmt) -> Option<Append<'a>> {
```

Comment:
For any methods here that only rely on `checker.semantic()`, I'd prefer to pass in `semantic: &'a SemanticModel` rather than the entire checker.

Reply:
No problem!